### PR TITLE
Causal attention

### PR DESCRIPTION
**Pull request by kirilklein**
## Summary by CodeRabbit

- **New Features**
  - Added a new configuration option for causal modeling in pretraining.
  - Introduced a constant for loss ignoring and a default vocabulary mapping for special tokens.
  - Added a function to create causal attention masks for sequence modeling.
  - Included attention mask tensor in dataset samples for improved model input handling.
  - Added detailed documentation for core model classes and methods.

- **Improvements**
  - Enhanced input validation and error handling in embedding modules.
  - Improved attention mask construction, supporting causal and sliding window masking.
  - Centralized ignore index for loss calculation to improve maintainability.
  - Expanded and clarified documentation for model components and configuration.
  - Refined model forward methods to support sparse prediction and optional inputs.

- **Packaging**
  - Updated package discovery settings for more robust builds.
  - Extended `.gitignore` to exclude additional directories.

- **Tests**
  - Added unit tests for embedding layers, model heads, and core model functionality.
  - Introduced tests for attention masking utilities to ensure correct behavior.

### CODE REVIEW

**Comment by kirilklein** [comment]
@coderabbitai summary

**Comment by kirilklein** [comment]
- option to get causal attention mask
- fix attention_mask generation
- cleanup introducing constants
- unittests for model module

File: corebehrt/modules/model/model.py [hunk 1]
```diff
@@ -27,23 +65,111 @@ def __init__(self, config):
...
+            (distance <= self.config.local_attention // 2)
+            .unsqueeze(0)
+            .unsqueeze(0)
+            .to(attention_mask.device)
+        )
+        # Combine with existing mask
+        sliding_window_mask = global_attention_mask.masked_fill(
+            window_mask.logical_not(), torch.finfo(self.dtype).min
+        )
+        if self.is_causal:
```
Comment (mikkelfo) [inline corebehrt/modules/model/model.py:159]: If the only change you want to add to our Model is this flag, then you could just do

```
def _update_attention_mask(...):
   global_attention_mask, sliding_window_mask = super()._update_attention_mask(...)
   global_attention_mask = make_attention_causal(global_attention_mask)
   sliding_window_mask = make_attention_causal(sliding_window_mask)
   return global_attention_mask, sliding_window_mask

**Review by mikkelfo — CHANGES_REQUESTED** [review]
Implementation wise it's fine (can improve with my comment). Still annoying we can't use FA2, but I guess this will do for now?

File: corebehrt/modules/model/model.py (hunk 1 above)
Comment (kirilklein) [reply corebehrt/modules/model/model.py:159]: good point

**Comment by kirilklein** [comment]
> Implementation wise it's fine (can improve with my comment). Still annoying we can't use FA2, but I guess this will do for now?

Yes, let's not worry about FA2 for now

**Review by mikkelfo — APPROVED** [review]
